tests/python/relay/test_pass_partition_graph.py

```diff
@@ -856,6 +857,111 @@ def expected():
     partitioned = transform.PartitionGraph()(mod)
     assert tvm.ir.structural_equal(partitioned, ref_mod, map_free_vars=True)
 
+
+def test_partition_conv_bias_relu():
+    def make_pattern():
+        data = relay.var("data", relay.TensorType((1, 3, 224, 224), "float32"))
+        weight = relay.var("weight")
+        bias = relay.var("bias")
+        conv = relay.nn.conv2d(data=data, weight=weight, kernel_size=(3, 3),
+                               channels=8, padding=(1, 1))
+        add = relay.add(conv, bias)
+        return relay.nn.relu(add)
+
+    def get_blocks(prefix, data, in_channel, out_channel,
+                   include_bn=True, include_sigmoid=False):
+        weight = relay.var(prefix + "weight")
+        bn_gamma = relay.var(prefix + "bn_gamma")
+        bn_beta = relay.var(prefix + "bn_beta")
+        bn_mmean = relay.var(prefix + "bn_mean")
+        bn_mvar = relay.var(prefix + "bn_var")
+
+        layer = relay.nn.conv2d(data=data, weight=weight, kernel_size=(3, 3),
+                                channels=out_channel, padding=(1, 1))
+        if include_bn:
+            bn_output = relay.nn.batch_norm(layer, bn_gamma, bn_beta,
+                                            bn_mmean, bn_mvar)
+            layer = bn_output[0]
+        if include_sigmoid:
+            # dummy layer to prevent pattern detection
+            layer = relay.sigmoid(layer)
+        layer = relay.nn.relu(layer)
+        return layer
+
+    def get_net(include_bn=True, include_sigmoid=False):
+        data = relay.var("data", relay.TensorType((1, 3, 224, 224), "float32"))
+        layer1 = get_blocks("layer1_", data, 3, 8, include_bn, include_sigmoid)
+        layer2 = get_blocks("layer2_", layer1, 8, 8, include_bn, include_sigmoid)
+        return relay.Function(relay.analysis.free_vars(layer2), layer2)
+
+    def get_partitoned_mod(mod, params):
+        # This is required for constant folding
+        mod["main"] = bind_params_by_name(mod["main"], params)
+        pattern_table = [
+            ("dnnl.conv_bias_relu", make_pattern())
+        ]
+        remove_bn_pass = transform.Sequential([
+            transform.InferType(),
+            transform.SimplifyInference(),
+            transform.FoldConstant(),
+            transform.FoldScaleAxis(),
+        ])
+        composite_partition = transform.Sequential([
+            remove_bn_pass,
+            transform.MergeComposite(pattern_table),
+            transform.AnnotateTarget("dnnl"),
```

Comment:
I'm just curious but is it possible to try MergeCompilerRegions here? In my own testing, I seem to be having issues with composite functions and that pass.

Reply:
I'm not familiar with what MergeCompilerRegions does, but assuming that it merges consecutive ops that are supported by the same compiler, yes it is possible in this case because all ops involved in the composite above, conv2d, add, and relu, are supported by the dnnl backend https://github.com/apache/incubator-tvm/blob/master/python/tvm/relay/op/contrib/dnnl.py#L59-L64. 

But I think composite is supposed to be more general, it works even if `_register_external_op_helper("add")` is missing in dnnl.py above, for example. The purpose of composite is to target a "complex" op whose ops that constitute it are not necessary supported by the compiler by itself.

Reply:
@alexwong Could you give some more detail on the issue you're facing? I'd be happy to take a look.

Reply:
@alexwong I'll add another pattern in the test, conv2d + relu. I can try MergeCompilerRegions + AnnotateTarget passes in that example. This way we can demonstrate both approaches.

Reply:
@mbaret If I choose MergeCompilerRegions + AnnotateTarget approach, can I still detect fused op inside codegen easily, like composite case?

Reply:
You can't. Part of the benefit of mergecomposite comes from wrapping the patterns in an easy-to-detect function.

Reply:
Thanks. @alexwong So the answer is both approaches are possible, but composite is more general and it also enables easily detecting fused ops inside codegen, like I do in this line https://github.com/apache/incubator-tvm/pull/5272/files#diff-1defdbd8a4c2ab55fb62ad44e9b314a8R256

Reply:
The motivation behind MergeCompilerRegions is when you have a codegen that wants to do its own graph-level optimisations, so it should see the full offloaded subgraph rather than one op at a time.

Reply:
> @alexwong Could you give some more detail on the issue you're facing? I'd be happy to take a look.

An example is I have a very simple Relay graph here where conv2d and clip are supported ops and squeeze and reshape are supported only if it's part of a squeeze->reshape pattern, so I am using mergecomposite for that.

```
def @main(%data: Tensor[(1, 3, 224, 224), float32], %weight1: Tensor[(32, 3, 3, 3), float32]) -> Tensor[(3, 3, 3, 32), float32] {
  %0 = nn.conv2d(%data, %weight1, padding=[1, 1, 1, 1], kernel_size=[3, 3]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %1 = clip(%0, a_min=0f, a_max=6f) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %2 = squeeze(%1) /* ty=Tensor[(32, 224, 224), float32] */;
  reshape(%2, newshape=[3, 3, 3, 32]) /* ty=Tensor[(3, 3, 3, 32), float32] */
}
```

After annotation I get:
```
def @main(%data: Tensor[(1, 3, 224, 224), float32], %weight1: Tensor[(32, 3, 3, 3), float32]) -> Tensor[(3, 3, 3, 32), float32] {
  %0 = annotation.compiler_begin(%data, meta[relay.attrs.CompilerAttrs][0]) /* ty=Tensor[(1, 3, 224, 224), float32] */;
  %1 = annotation.compiler_begin(%weight1, meta[relay.attrs.CompilerAttrs][1]) /* ty=Tensor[(32, 3, 3, 3), float32] */;
  %2 = nn.conv2d(%0, %1, padding=[1, 1, 1, 1], kernel_size=[3, 3]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %3 = annotation.compiler_end(%2, meta[relay.attrs.CompilerAttrs][2]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %4 = annotation.compiler_begin(%3, meta[relay.attrs.CompilerAttrs][3]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %5 = clip(%4, a_min=0f, a_max=6f) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %6 = annotation.compiler_end(%5, meta[relay.attrs.CompilerAttrs][4]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %7 = annotation.compiler_begin(%6, meta[relay.attrs.CompilerAttrs][5]) /* ty=Tensor[(1, 32, 224, 224), float32] */;
  %9 = fn (%x: Tensor[(1, 32, 224, 224), float32], Composite="compilername.squeeze") -> Tensor[(3, 3, 3, 32), float32] {
    %8 = squeeze(%x) /* ty=Tensor[(32, 224, 224), float32] */;
    reshape(%8, newshape=[3, 3, 3, 32]) /* ty=Tensor[(3, 3, 3, 32), float32] */
  };
  %10 = %9(%7) /* ty=Tensor[(3, 3, 3, 32), float32] */;
  annotation.compiler_end(%10, meta[relay.attrs.CompilerAttrs][6]) /* ty=Tensor[(3, 3, 3, 32), float32] */
}
```

This looks right to me but mergecompilerregions is throwing the following error: 
```
tvm._ffi.base.TVMError: TVMError: Cannot find the corresponding region for start annotation:
v0.0.4
free_var %x: Tensor[(1, 32, 224, 224), float32]
%0 = annotation.compiler_begin(%x, meta[relay.attrs.CompilerAttrs][0]);
%1 = squeeze(%0);
%2 = annotation.compiler_end(%1, meta[relay.attrs.CompilerAttrs][1]);
annotation.compiler_begin(%2, meta[relay.attrs.CompilerAttrs][2])
// meta data omitted. you can use show_meta_data=True to include meta data
```

Not sure if it's expected to not work with composite functions but I think there may be some missing logic there.

Reply:
Yep, that looks like a bug :) I'll take a quick look now, but I'll probably have a fix PR up tomorrow.

Reply:
That'd be great! Let me know your initial findings, might be something small :)

Reply:
Any updates / initial findings?

Reply:
Hey I just tried your case with #5277 and it works fine. Maybe we can push to merge that PR so that this problem will be gone naturally.

Reply:
Looks like that is caused by wrongly adding the `default` compiler annotations. Let's work on reviewing #5277 and merge these bunch of fixes.

Reply:
Yes, this PR fixed it!